sublimeMerge: init at 1058

jonoabroad opened the pull request:
###### Motivation for this change
Provide git client Sublime Merge  addressing https://github.com/NixOS/nixpkgs/issues/47377
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
Sublime Merge: Initial commit, version 1058

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18806,6 +18806,10 @@ with pkgs;
 
+  sublimeMergePackages = recurseIntoAttrs (callPackage ../development/tools/sublime-merge/packages.nix { });
+
+  sublimeMerge = sublimeMergePackages.sublimeMerge;
```
Why not use `inherit`?

jonoabroad replied:
fixing other comments, I'll explain this one though - this is a copy of the sublime text 3 package with as minimal changes as necessary to get it working.

jtojnar commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,90 @@
+    homepage = https://www.sublimemerge.com/;
+    maintainers = with maintainers; [ jonoabroad ];
+    license = licenses.unfree;
+    platforms = [ "x86_64-linux" "i686-linux" ];
```
There are only 64 bit binaries.

jtojnar commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,90 @@
+      done
+
+      # Rewrite pkexec|gksudo argument. Note that we can't delete bytes in binary.
+      sed -i -e 's,/bin/cp\x00,cp\x00\x00\x00\x00\x00\x00,g' sublime_merge
```
This does not seem to be necessary.

jtojnar commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,90 @@
+  installPhase = ''
+    mkdir -p $out/bin
+
+    cat > $out/bin/sublM <<-EOF
```
The short name is `smerge` not `sublM`, see the debian package.

jtojnar commented on pkgs/development/tools/sublime-merge/packages.nix:
```diff
@@ -0,0 +1,11 @@
+      buildVersion = "1058";
+      x64sha256 = "1yj3cg3z9421h36mgjprr94p4m0i4hl4ndi215bfp1i1b02rrva8";
+    } {};
+  }
```
Missing `\n` on the last line.

jtojnar commented:
Could you add dev builds as well? https://forum.sublimetext.com/t/sublime-merge-dev-build-1060/39348

Commit:
Tidying based on feedback:
 - short name smerge rather than sublM,
 - only support 64bit,
 - add current dev build,
 - remove unneeded rewrite.

Mic92 commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,87 @@
+      # when trying to run a build from within Sublime Merge
+      ln -s ${bash}/bin/bash $out/sublime_bash
+      wrapProgram $out/sublime_bash \
+        --set LD_PRELOAD "${stdenv.cc.cc.lib}/lib${stdenv.lib.optionalString stdenv.is64bit "64"}/libgcc_s.so.1"
```
This is weird so. Does it not like `libredirect`? You could use `makeWrapper` to create `sublime_bash` directly instead of having the symlink first.

Mic92 commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,87 @@
+    #!/bin/sh
+    exec $sublimeMerge/sublime_merge "\$@"
+    EOF
+    chmod +x $out/bin/smerge
```
This is a job for `makeWrapper`.

Mic92 commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,87 @@
+    chmod +x $out/bin/smerge
+
+    ln $out/bin/smerge $out/bin/sublime_merge
+    ln $out/bin/smerge $out/bin/sublimeMerge
```
Why not just wrap to `$out/bin/sublime_merge` directly and then do the symlink?

Mic92 commented on pkgs/development/tools/sublime-merge/common.nix:
```diff
@@ -0,0 +1,87 @@
+
+  libPath = stdenv.lib.makeLibraryPath [glib xorg.libX11 gtk2 cairo pango];
+  redirects = [ "/usr/bin/pkexec=${pkexecPath}" ]
+    ++ stdenv.lib.optional gksuSupport "/usr/bin/gksudo=${gksu}/bin/gksudo";
```
Maybe we can convince upstream to not use absolute paths here.

Mic92 replied:
Is the sublime forum the right place to ask? @guillermooo

Mic92 replied:
I dropped something here, let's see: https://forum.sublimetext.com/t/sublime-tries-absolute-paths-for-pkexec-gksudo/39362

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18786,6 +18786,13 @@ with pkgs;
 
+  sublimeMergePackages = recurseIntoAttrs (callPackage ../development/tools/sublime-merge/packages.nix { });
+
+  sublimeMerge = sublimeMergePackages.sublimeMerge;
```
You can do `inherit (sublimeMergePackages) sublimeMerge sublimeMerge-dev;`

jonoabroad replied:
Oh, that is much nicer.

ryantm reviewed (changes requested):
Please address the other review comments.

jtojnar commented:
Sublime Text now has `updateScript`, it would be nice to add it here as well.

worldofpeace commented:
closing in favor of #58994
